project-koku/koku
# Allowing koku provider and platform-sources to run simultaneously

This change will enable us to run our existing provider workflow in addition to adding/deleting sources via Platform-Sources.

If a provider already exists in koku and an identical source is added via Sources; Koku API will detect this and update the Sources database table with the correct koku UUID.

Delete checks were also added so that Providers added via Platform-Sources must be deleted by Platform-Sources.  Likewise, providers added directly must be deleted directly. 

**Testing**
1. Add a Source via Sources API and attempt to remove from koku directly.
2.  Attempt to delete provider from step 1 directly.  Verify error is thrown
3. Delete provider via Platform-Sources and verify delete was successful.
4.  Add Koku provider directly
5.  Add same provider via Sources and verify that "add" is successful and that the koku_uuid field is updated on the Sources table.
6.  Verify that the provider added from step 4 can only be deleted via Sources now and is blocked when deleting directly.

**Test Results**
[sources_coexist_ut.txt](https://github.com/project-koku/koku/files/3617657/sources_coexist_ut.txt)
